File: server/src/cyberq.go
package gocyberq

import (
        // "encoding/json"
        "encoding/xml"
        // "net/http"
        "fmt"
        "io/ioutil"
        "os"
)

type Nutcstatus struct {
     OutputPercent string `xml:"OUTPUT_PERCENT"`
     TimerCurrent string `xml:"TIMER_CURR"`
     CookTemp string `xml:"COOK_TEMP"`
     Food1Temp string `xml:"FOOD1_TEMP"`
     Food2Temp string `xml:"FOOD2_TEMP"`
     Food3Temp string `xml:"FOOD3_TEMP"`
     CookStatus string `xml:"COOK_STATUS"`
     Food1Status string `xml:"FOOD1_STATUS"`
     Food2Status string `xml:"FOOD2_STATUS"`
     Food3Status string `xml:"FOOD3_STATUS"`
     TimerStatus string `xml:"TIMER_STATUS"`
     DegUnits string `xml:"DEG_UNITS"`
     CookCycleTime string `xml:"COOK_CYCTIME"`
     CookPropBand string `xml:"COOK_PROPBAND"`
     CookRamp string `xml:"COOK_RAMP"`
}

type Nutcallstatus struct {
    Cook Cook `xml:"COOK"`
    Food1 Food1 `xml:"FOOD1"`
    Food2 Food2 `xml:"FOOD2"`
    Food3 Food3 `xml:"FOOD3"`
    OutputPercent string `xml:"OUTPUT_PERCENT"`
    TimerCurrent string `xml:"TIMER_CURR"`
    TimerStatus string `xml:"TIMER_STATUS"`
    DegUnits string `xml:"DEG_UNITS"`
    CookCycleTime string `xml:"COOK_CYCTIME"`
    CookPropBand string `xml:"COOK_PROPBAND"`
    CookRamp string `xml:"COOK_RAMP"`

    // Used only in the config call
    System System `xml:"SYSTEM"`
    Control Control `xml:"CONTROL"`
}

type Cook struct {
    Name string `xml:"COOK_NAME"`
    Temp string `xml:"COOK_TEMP"`
    Set string `xml:"COOK_SET"`
    Status string `xml:"COOK_STATUS"`
}

type Food1 struct {
    Name string `xml:"FOOD1_NAME"`
    Temp string `xml:"FOOD1_TEMP"`
    Set string `xml:"FOOD1_SET"`
    Status string `xml:"FOOD1_STATUS"`
}

type Food2 struct {
    Name string `xml:"FOOD2_NAME"`
    Temp string `xml:"FOOD2_TEMP"`
    Set string `xml:"FOOD2_SET"`
    Status string `xml:"FOOD2_STATUS"`
}

type Food3 struct {
    Name string `xml:"FOOD3_NAME"`
    Temp string `xml:"FOOD3_TEMP"`
    Set string `xml:"FOOD3_SET"`
    Status string `xml:"FOOD3_STATUS"`
}

type System struct {
    MenuScrolling string `xml:"MENU_SCROLLING"`
    LcdBacklight string `xml:"LCD_BACKLIGHT"`
    LcdContrast string `xml:"LCD_CONTRAST"`
    DegUnits string `xml:"DEG_UNITS"`
    AlarmBeeps string `xml:"ALARM_BEEPS"`
    KeyBeeps string `xml:"KEY_BEEPS"`
}

type Control struct {
    TimeoutAction string `xml:"TIMEOUT_ACTION"`
    CookHold string `xml:"COOKHOLD"`
    AlarmDev string `xml:"ALARMDEV"`
    CookRamp string `xml:"COOK_RAMP"`
    OpenDetect string `xml:"OPENDETECT"`
    CycleTime string `xml:"CYCTIME"`
    PropBand string `xml:"PROPBAND"`
}

func Status() (status Nutcstatus) {
    xmlFile, err := os.Open("../docs/cyberq_status.xml")
    if err != nil {
        fmt.Println("Error opening file:", err)
        return
    }
    defer xmlFile.Close()

    b, _ := ioutil.ReadAll(xmlFile)

    xml.Unmarshal(b, &status)

    // fmt.Println(status)
    return status
}

func All() (status Nutcallstatus) {
    xmlFile, err := os.Open("../docs/cyberq_all.xml")
    if err != nil {
        fmt.Println("Error opening file:", err)
        return
    }
    defer xmlFile.Close()

    b, _ := ioutil.ReadAll(xmlFile)

    xml.Unmarshal(b, &status)

    // fmt.Println(status)
    return status
}

func Config() (status Nutcallstatus) {
    xmlFile, err := os.Open("../docs/cyberq_config.xml")
    if err != nil {
        fmt.Println("Error opening file:", err)
        return
    }
    defer xmlFile.Close()

    b, _ := ioutil.ReadAll(xmlFile)

    xml.Unmarshal(b, &status)

    // fmt.Println(status)
    return status
}

File: node/cyberq.js
"use strict";

var Agent = require('superagent-promise')(require('superagent'), Promise);
var Util = require('util')
var xml2js = require('xml2js');
var _ = require('lodash');

class CyberQ {
	constructor(config) {
		this.host = config.cyberq.host;
	}

	current() {
		return Agent.get(this.getRequestPath('all.xml'))
				.then((response) => {
					return new Promise((resolve, reject) => {
						xml2js.parseString(response.text, (err, result) => {
							if (!err) {
								resolve(result);
							} else {
								reject(err);
							}
						});
					});
				})
				// Log for debugging request reponses.
				// .then((result) => {
				// 	console.log(Util.inspect(result, {depth: null, colors: true}));
				// 	return result;
				// })
				.then((result) => {
					return new CyberQStatus(result);
				});
	}

	getRequestPath(path) {
		return `http://${this.host}/${path}`;
	}
}
/*
{ nutcallstatus:
   { COOK:
      [ { COOK_NAME: [ 'Big Green Egg' ],
          COOK_TEMP: [ '3216' ],
          COOK_SET: [ '4000' ],
          COOK_STATUS: [ '0' ] } ],
     FOOD1:
      [ { FOOD1_NAME: [ 'Chicken Quarters' ],
          FOOD1_TEMP: [ '1482' ],
          FOOD1_SET: [ '1750' ],
          FOOD1_STATUS: [ '0' ] } ],
     FOOD2:
      [ { FOOD2_NAME: [ 'Food2' ],
          FOOD2_TEMP: [ 'OPEN' ],
          FOOD2_SET: [ '1000' ],
          FOOD2_STATUS: [ '4' ] } ],
     FOOD3:
      [ { FOOD3_NAME: [ 'Food3' ],
          FOOD3_TEMP: [ 'OPEN' ],
          FOOD3_SET: [ '1000' ],
          FOOD3_STATUS: [ '4' ] } ],
     OUTPUT_PERCENT: [ '100' ],
     TIMER_CURR: [ '00:00:00' ],
     TIMER_STATUS: [ '0' ],
     DEG_UNITS: [ '1' ],
     COOK_CYCTIME: [ '6' ],
     COOK_PROPBAND: [ '500' ],
     COOK_RAMP: [ '0' ] } }
*/
class CyberQStatus {
	constructor(statusJson) {
		this.pit = new CyberQProbStatus(statusJson.nutcallstatus.COOK);
		this.food1 = new CyberQProbStatus(statusJson.nutcallstatus.FOOD1);
		this.food2 = new CyberQProbStatus(statusJson.nutcallstatus.FOOD2);
		this.food3 = new CyberQProbStatus(statusJson.nutcallstatus.FOOD3);
		this.output = parseInt(statusJson.nutcallstatus.OUTPUT_PERCENT[0]);
		this.timestamp = Date.now();
	}

	pit() { return this.pit; }
	food1() { return this.food1; }
	food2() { return this.food2; }
	food3() { return this.food3; }
	output() { return this.output; }
}

class CyberQProbStatus {
	constructor(statusJson) {
		if (_.isArray(statusJson)) {
			statusJson = statusJson[0];
		}

		for (var key in statusJson) {
			if (/NAME/.test(key)) {
				this.name = statusJson[key][0];
			} else if (/TEMP/.test(key)) {
				this.temp = parseInt(statusJson[key][0]);
			} else if (/SET/.test(key)) {
				this.set = parseInt(statusJson[key][0]);
			} else if (/STATUS/.test(key)) {
				this.status = parseInt(statusJson[key][0]);
			}
		}
	}

	name() { return this.name; }
	temp() { return this.temp; }
	set() { return this.set; }
	status() { return this.status; }
}

module.exports = CyberQ;
